source/server/listener_manager_impl.cc

```diff
@@ -88,19 +90,43 @@ ListenerImpl::ListenerImpl(const envoy::api::v2::Listener& config, ListenerManag
       local_drain_manager_(parent.factory_.createDrainManager()) {
   // TODO(htuch): Support multiple filter chains #1280, add constraint to ensure we have at least on
   // filter chain #1308.
-  ASSERT(config.filter_chains().size() == 1);
-  const auto& filter_chain = config.filter_chains()[0];
-
-  listener_scope_ =
-      parent_.server_.stats().createScope(fmt::format("listener.{}.", address_->asString()));
+  ASSERT(config.filter_chains().size() >= 1);
+
+  // Skip lookup and update of the SSL Context if there is only one filter chain
+  // and it doesn't enforce any SNI restrictions.
+  bool skip_context_update;
+  if (config.filter_chains().size() == 1 &&
+      config.filter_chains()[0].filter_chain_match().sni_domains().empty()) {
+    skip_context_update = true;
+  } else {
+    skip_context_update = false;
+  }
 
-  if (filter_chain.has_tls_context()) {
-    Ssl::ServerContextConfigImpl context_config(filter_chain.tls_context());
-    ssl_context_ = parent_.server_.sslContextManager().createSslServerContext(*listener_scope_,
-                                                                              context_config);
+  size_t has_tls = 0;
+  size_t has_stk = 0;
+  for (const auto& filter_chain : config.filter_chains()) {
+    std::vector<std::string> sni_domains(filter_chain.filter_chain_match().sni_domains().begin(),
+                                         filter_chain.filter_chain_match().sni_domains().end());
+    if (filter_factories_.empty() || !filter_chain.filters().empty()) {
+      // Only one filter chain can have filters until multiple filter chains are properly supported.
+      RELEASE_ASSERT(filter_factories_.empty());
+      filter_factories_ = parent_.factory_.createFilterFactoryList(filter_chain.filters(), *this);
+    }
+    if (filter_chain.has_tls_context()) {
+      Ssl::ServerContextConfigImpl context_config(filter_chain.tls_context());
+      tls_contexts_.emplace_back(parent_.server_.sslContextManager().createSslServerContext(
+          name_, sni_domains, *listener_scope_, context_config, skip_context_update));
+      has_tls++;
+      if (filter_chain.tls_context().has_session_ticket_keys()) {
+        has_stk++;
+      }
+    }
   }
 
-  filter_factories_ = parent_.factory_.createFilterFactoryList(filter_chain.filters(), *this);
+  // TODO(PiotrSikora): allow filter chains with mixed use of Session Ticket Keys.
```

Comment:
Can you add a comment on why this won't work as is?  I thought this would have worked as-is.

Reply:
Done.

Note that there are few ways to make it work, but I don't want to block this PR on this.